twenty-first/src/util_types/merkle_tree.rs

```diff
@@ -1033,4 +1042,166 @@ pub mod merkle_tree_test {
 
         assert_eq!(expected_paths, auth_paths);
     }
+
+    fn root_from_authentication_struct(
+        auth_struct: Vec<Digest>,
+        auth_struct_indices: Vec<u64>,
+        leafs: Vec<Digest>,
+        leafs_node_indices: Vec<u64>,
+        nd_siblings: Vec<(Digest, Digest)>,
+        nd_indices: Vec<(u64, u64)>,
+    ) -> Digest {
+        fn digest_to_xfe(digest: Digest, challenge: XFieldElement) -> XFieldElement {
+            let leaf_xfe_lo = XFieldElement::new([digest.0[0], digest.0[1], digest.0[2]]);
+            let leaf_xfe_hi =
+                challenge * XFieldElement::new([digest.0[3], digest.0[4], BFieldElement::one()]);
+
+            leaf_xfe_lo + leaf_xfe_hi
+        }
+
+        assert_eq!(leafs.len(), leafs_node_indices.len());
+        assert_eq!(auth_struct.len(), auth_struct_indices.len());
+        let (alpha, gamma, delta) = {
+            let mut sponge = Tip5::init();
+            sponge.pad_and_absorb_all(&leafs.encode());
+            sponge.pad_and_absorb_all(&auth_struct.encode());
+            let challenges = sponge.sample_scalars(3);
+            (challenges[0], challenges[1], challenges[2])
+        };
+
+        let leafs_as_xfes = leafs.iter().map(|l| digest_to_xfe(*l, alpha)).collect_vec();
+        let auth_str_as_xfes = auth_struct
+            .iter()
+            .map(|l| digest_to_xfe(*l, alpha))
+            .collect_vec();
+
+        let mut p = XFieldElement::one();
+        for i in 0..leafs.len() {
+            let leaf_index_hi = bfe!(leafs_node_indices[i] >> 32);
+            let leaf_index_lo = bfe!(leafs_node_indices[i] & u32::MAX as u64);
+            let leaf_index_xfe =
+                XFieldElement::new([leaf_index_hi, leaf_index_lo, BFieldElement::zero()]);
+            let fact = leafs_as_xfes[i] - gamma + delta * leaf_index_xfe;
+            p *= fact;
+        }
+        for i in 0..auth_struct.len() {
+            let leaf_index_hi = bfe!(auth_struct_indices[i] >> 32);
+            let leaf_index_lo = bfe!(auth_struct_indices[i] & u32::MAX as u64);
+            let leaf_index_xfe =
+                XFieldElement::new([leaf_index_hi, leaf_index_lo, BFieldElement::zero()]);
```

Comment:
Why are you putting the more significant limb (`hi`) on the least significant coefficient, but nothing on the most significant coefficient? It's counter-intuitive.